internal/webhooks/cluster.go

```diff
@@ -395,6 +390,229 @@ func (webhook *Cluster) validateTopology(ctx context.Context, oldCluster, newClu
+		Patch: 0,
+	}
+
+	// Return early if its not a minor version upgrade.
+	if !inVersion.GTE(ceilVersion) {
+		return nil
+	}
+
+	allErrs := field.ErrorList{}
+	// minor version cannot be increased if KCP is upgrading or not yet on the current version
+	if err := validateTopologyControlPlaneVersion(ctx, fldPath, webhook.Client, oldCluster, oldVersion); err != nil {
+		allErrs = append(allErrs, field.Invalid(
+			fldPath,
+			fldValue,
+			fmt.Sprintf("blocking version update due to ControlPlane version check: %v", err),
+		))
+	}
+
+	// minor version cannot be increased if MachineDeployments are upgrading or not yet on the current version
+	if err := validateTopologyMachineDeploymentVersions(ctx, fldPath, webhook.Client, oldCluster, oldVersion); err != nil {
+		allErrs = append(allErrs, field.Invalid(
+			fldPath,
+			fldValue,
+			fmt.Sprintf("blocking version update due to MachineDeployment version check: %v", err),
+		))
+	}
+
+	// minor version cannot be increased if MachinePools are upgrading or not yet on the current version
+	if err := validateTopologyMachinePoolVersions(ctx, fldPath, webhook.Client, webhook.Tracker, oldCluster, oldVersion); err != nil {
+		allErrs = append(allErrs, field.Invalid(
+			fldPath,
+			fldValue,
+			fmt.Sprintf("blocking version update due to MachinePool version check: %v", err),
+		))
+	}
+
+	if len(allErrs) > 0 {
+		return allErrs
+	}
+
+	return nil
+}
+
+func validateTopologyControlPlaneVersion(ctx context.Context, fldPath *field.Path, ctrlClient client.Reader, oldCluster *clusterv1.Cluster, oldVersion semver.Version) error {
+	cp, err := external.Get(ctx, ctrlClient, oldCluster.Spec.ControlPlaneRef, oldCluster.Namespace)
+	if err != nil {
+		return field.InternalError(fldPath, err)
+	}
+
+	cpVersionString, err := contract.ControlPlane().Version().Get(cp)
+	if err != nil {
+		return field.InternalError(fldPath, err)
+	}
+
+	cpVersion, err := semver.ParseTolerant(*cpVersionString)
+	if err != nil {
+		// NOTE: this should never happen. Nevertheless, handling this for extra caution.
+		return field.InternalError(fldPath, errors.New("unable to parse version of ControlPlane"))
+	}
+	if cpVersion.NE(oldVersion) {
+		return fmt.Errorf("ControlPlane version %q does not match the current version %q", cpVersion, oldVersion)
+	}
+
+	provisioning, err := contract.ControlPlane().IsProvisioning(cp)
+	if err != nil {
+		return field.InternalError(fldPath, err)
+	}
+
+	if provisioning {
+		return errors.New("ControlPlane is currently provisioning")
+	}
+
+	upgrading, err := contract.ControlPlane().IsUpgrading(cp)
+	if err != nil {
+		return field.InternalError(fldPath, err)
+	}
+
+	if upgrading {
+		return errors.New("ControlPlane is currently upgrading")
+	}
+
+	return nil
+}
+
+func validateTopologyMachineDeploymentVersions(ctx context.Context, fldPath *field.Path, ctrlClient client.Reader, oldCluster *clusterv1.Cluster, oldVersion semver.Version) error {
+	// FROM: desired_state.go getCurrentMachineDeploymentState
```

Comment:
```suggestion
	// FROM: current_state.go getCurrentMachineDeploymentState
```
Sorry, my bad in the previous suggestion (same below for MP)

Reply:
Maybe let's move this comment after l.509. I think now basically only the List command is the same